### PR TITLE
Fix Staking Accuracy issues.

### PR DESCRIPTION
**Pull request by Kianenigma**
- Small patch for recent staking issues in test-net.

### CODE REVIEW

File: srml/staking/src/phragmen.rs [hunk 1]
```diff
@@ -24,6 +24,7 @@ use crate::{Exposure, BalanceOf, Trait, ValidatorPrefs, IndividualExposure};
 
 type Fraction = PerU128;
 type ExtendedBalance = u128;
+const SCALE_FACTOR_64: u128 = u64::max_value() as u128;
```
Comment (gavofyork) [inline srml/staking/src/phragmen.rs]: if you make this `u64::max_value() as u128 + 1`, then it'll become a simple shift left and shift right in the code which should be a lot faster in wasm.

File: srml/staking/src/phragmen.rs [hunk 2]
```diff
@@ -214,8 +220,7 @@ pub fn elect<T: Trait + 'static, FV, FN, FS>(
 				// if the target of this vote is among the winners, otherwise let go.
 				if let Some(c) = elected_candidates.iter_mut().find(|c| c.who == e.who) {
 					e.elected = true;
-					// NOTE: for now, always divide last to avoid collapse to zero.
-					e.backing_stake = n.budget.saturating_mul(*e.load) / n.load.max(1);
+					e.backing_stake = n.budget.saturating_mul(*e.load) / (*n.load).max(1);
```
Comment (gavofyork) [inline srml/staking/src/phragmen.rs]: are these parens really needed?

File: srml/staking/src/phragmen.rs (hunk 1 above)
Comment (gavofyork) [reply srml/staking/src/phragmen.rs]:
```suggestion
const SCALE_FACTOR_64: u128 = (u64::max_value() as u128) + 1;
```

File: srml/staking/src/phragmen.rs (hunk 2 above)
Comment (gavofyork) [reply srml/staking/src/phragmen.rs]:
```suggestion
					e.backing_stake = n.budget.saturating_mul(*e.load) / n.load.max(1);
```

**Comment by gavofyork** [comment]
Will need a runtime version bump, too.

File: srml/staking/src/phragmen.rs (hunk 2 above)
Comment (Kianenigma) [reply srml/staking/src/phragmen.rs]: No, and this particular one does not serve any purpose frankly. Some other ones were there the emphasize the order of operations (left-associativity already works fine). All removed and replaced with comments to note the importance when needed.

File: srml/staking/src/phragmen.rs [hunk 3]
```diff
@@ -178,7 +179,12 @@ pub fn elect<T: Trait + 'static, FV, FN, FS>(
 				for e in &n.edges {
 					let c = &mut candidates[e.candidate_index];
 					if !c.elected && !c.approval_stake.is_zero() {
-						let temp = n.budget.saturating_mul(*n.load) / c.approval_stake;
+						let temp =
+							// Basic fixed-point shifting by 64.
+							// This will never saturate since n.budget cannot exceed u64,despite being stored in u128.
+							// Note that left-associativity in operators precedence is crucially important here.
+							n.budget.saturating_mul(SCALE_FACTOR_64) / c.approval_stake
+							* (*n.load / SCALE_FACTOR_64);
```
Comment (rphmeier) [inline srml/staking/src/phragmen.rs]: would it not be more accurate to divide the whole thing by `SCALE_FACTOR_64` in the end, as it's likely to be a larger number?

File: srml/staking/src/tests.rs [hunk 4]
```diff
@@ -1959,3 +1959,105 @@ fn phragmen_should_not_overflow_ultimate() {
 		check_exposure(2);
 	})
 }
+
+
+#[test]
+fn large_scale_test() {
+	with_externalities(&mut ExtBuilder::default()
+	.nominate(false)
```
Comment (rphmeier) [inline srml/staking/src/tests.rs]: indent!

File: srml/staking/src/phragmen.rs (hunk 3 above)
Comment (rphmeier) [reply srml/staking/src/phragmen.rs]: also, this line needs another level of indent

**Pull request merged**